File: swift/0706-design-hashmap.swift
/**
 * Question Link: https://leetcode.com/problems/design-hashmap/
 */

 class Pair {
    let key: Int
    var val: Int

    init(_ key: Int, _ val: Int) {
        self.key = key
        self.val = val
    }
}

class MyHashMap {
    var size: Int
    var capacity: Int
    var map: [Pair?]

    init() {
        self.size = 0
        self.capacity = 10 * 10 * 10 * 10 * 10
        self.map = [Pair?](repeating: nil, count: capacity)
    }

    func hash(_ key: Int) -> Int {
        return key % capacity
    }

    func rehash() {
        capacity = 2 * capacity
        var newMap = [Pair?](repeating: nil, count: capacity)
        let oldMap = map
        map = newMap
        for pair in oldMap {
            if pair != nil {
                map.append(pair)
            }
        }
    }
    
    func put(_ key: Int, _ value: Int) {
        var index = hash(key)

        while true {
            if map[index] == nil {
                map[index] = Pair(key, value)
                size += 1
                if size >= capacity / 2 {
                    rehash()
                }
                return
            } else if map[index]?.key == key {
                map[index]?.val = value
                return
            }
            index += 1
            index = index % capacity
        }
    }
    
    func get(_ key: Int) -> Int {
        var index = hash(key)

        while map[index] != nil {
            if map[index]!.key == key {
                return map[index]!.val
            }
            index += 1
            index = index % capacity
        }

        return -1
    }
    
    func remove(_ key: Int) {
        if get(key) == -1 {
            return
        }

        var index = hash(key)
        while true {
            if map[index]?.key == key {
                map[index] = nil
                size -= 1
                return
            }
            index += 1
            index = index % capacity
        }
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * let obj = MyHashMap()
 * obj.put(key, value)
 * let ret_2: Int = obj.get(key)
 * obj.remove(key)
 */

File: kotlin/0706-design-hashmap.kt
class ChainNode(
    var key: Int = -1,
    var value: Int = -1
) {
    var next: ChainNode? = null
}

class MyHashMap() {

    val hashMap = Array(1000) { ChainNode() }

    fun put(key: Int, value: Int) {
        var current: ChainNode? = hashMap[key % hashMap.size]
        while(current?.next != null) {
            if(current.next?.key == key) {
                current.next?.value = value
                return
            }
            current = current?.next
        }
        current?.next = ChainNode(key, value)
    }

    fun get(key: Int): Int {
        var current: ChainNode? = hashMap[key % hashMap.size].next
        while(current != null) {
            if(current.key == key) return current.value
            current = current.next
        }
        return -1
    }

    fun remove(key: Int) {
        var current: ChainNode? = hashMap[key % hashMap.size]
        while(current != null && current.next != null) {
            if(current.next?.key == key) {
                current.next = current.next?.next
                return
            }
            current = current.next
        }
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * var obj = MyHashMap()
 * obj.put(key,value)
 * var param_2 = obj.get(key)
 * obj.remove(key)
 */
